### PR TITLE
Schema model improvements

### PR DESCRIPTION
**Pull request by angrykoala**
# Description
This PR improves on the schema model by adding `Relationship` to it, as well as some extra data to `Attribute`.

This is a first step required to move to a full schema-model translation layer

## Complexity

Complexity: Medium

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 43 | 39 | 1134 |
|  aggregations.NestedAggregation | 14551 | 14551 | 106 | 97 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 70 | 89 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 2866 | 3130 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 2827 | 2890 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 55 | 57 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 57 | 131 | 2174 |
|  aggregations.DeeplyNestedCount | 13070327 | 13070333 | 4335 | 4495 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 150 | 186 | 600 |
|  connect.createAndConnect | 14424 | 14424 | 182 | 167 | 3003 |
|  connections.Connection | 14082 | 14082 | 86 | 96 | 2174 |
|  connections.NestedConnection | 41462 | 41464 | 101 | 146 | 4516 |
|  connections.ConnectionWithSort | 2282 | 2282 | 55 | 63 | 1040 |
|  connections.ConnectionWithSortAndCypher | 2282 | 2282 | 155 | 244 | 1040 |
|  create.SimpleMutation | 7 | 7 | 51 | 73 | 1 |
|  cypher-directive.TopLevelCypherDirective | 116092 | 116092 | 689 | 782 | 12241 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 147611 | 147611 | 313 | 314 | 12241 |
|  delete.SimpleDelete | 19401 | 19401 | 329 | 269 | 1040 |
|  delete.NestedDeleteInUpdate | 22200 | 22200 | 129 | 192 | 2040 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 34 | 47 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 74 | 102 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 3450 | 3450 | 22 | 25 | 1134 |
|  2925.NestedSingleRelationshipRequiredFilter | 9383 | 9383 | 50 | 48 | 1040 |
|  query.SimpleQuery | 15160 | 15160 | 55 | 86 | 2174 |
|  query.QueryWhere | 9705 | 9696 | 30 | 36 | 2164 |
|  query.SimpleQueryWithNestedWhere | 9883 | 9874 | 44 | 51 | 2164 |
|  query.Nested | 10092037 | 10092037 | 10454 | 10725 | 2008534 |
|  query.NestedWithFilter | 40401 | 40401 | 95 | 139 | 4511 |
|  query.OrFilterOnRelationships | 43044 | 40826 | 144 | 180 | 1706 |
|  query.OrFilterOnRelationshipsAndNested | 36558 | 35196 | 211 | 247 | 1706 |
|  query.QueryWithNestedIn | 14266 | 13393 | 55 | 59 | 1507 |
|  query.NestedConnectionWhere | 9834 | 9834 | 55 | 59 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9939 | 10059 | 66 | 94 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 19194 | 19308 | 125 | 138 | 1614 |
|  query.NestedWithRelationshipSingleFilters | 3024400 | 3024400 | 580 | 590 | 1003942 |
|  query.Fulltext | 96 | 96 | 23 | 26 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 37 | 41 | 84 |
|  sorting.SortMultipleTypes | 2513 | 2513 | 60 | 69 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1485 | 1434 | 172 | 212 | 1040 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 76 | 103 | 2174 |
|  sorting.SortDeeplyNestedFields | 42196 | 42196 | 140 | 120 | 4516 |
|  sorting.SortWithTopLevelCypher | 2119 | 2119 | 94 | 153 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 71 | 72 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 53 | 63 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 322 | 323 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 296 | 328 | 33033 |
|  update.NestedUpdate | 2119 | 2119 | 58 | 77 | 1040 |

</details>
        
Old Schema Generation: 27.679s
Schema Generation: 28.277s
Old Subgraph Schema Generation: 42.437s
Subgraph Schema Generation: 41.277s

File: packages/graphql/src/schema-model/relationship/Relationship.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+export class Relationship {
+    public readonly name: string;
+    public readonly type: string;
+    public readonly attributes: Map<string, Attribute> = new Map();
+    public readonly source: ConcreteEntity; // Origin field of relationship
+    public readonly target: Entity;
+    public readonly direction: RelationshipDirection;
+
+    /**Note: Required for now to infer the types without ResolveTree */
+    public get connectionFieldTypename(): string {
```
Comment (MacondoExpress) [inline packages/graphql/src/schema-model/relationship/Relationship.ts:37]: This is interesting, but I'm not 100% sure about it, do we want to start adding a dependency of the augmentedSchema generated types in the schema model? Not as a blocker but as a conversation point for the whole IR design!

File: packages/graphql/src/schema-model/generate-model.ts [hunk 2]
```diff
@@ -153,15 +239,27 @@ function getLabels(definition: ObjectTypeDefinitionNode, nodeDirectiveArguments:
...
         return new Attribute({
             name: field.name.value,
             annotations,
+            type: typeMeta.name,
+            isArray: Boolean(typeMeta.array),
         });
     }
 }
 
+function isAttributeType(typeName: string): typeName is AttributeType {
```
Comment (MacondoExpress) [inline packages/graphql/src/schema-model/generate-model.ts:253]: Potentially we can start moving these utility functions in some parser utils file.

File: packages/graphql/src/schema-model/generate-model.ts [hunk 3]
```diff
@@ -121,8 +131,84 @@ function generateCompositeEntity(
...
+): void {
+    const name = definition.name.value;
+    const entity = schema.getEntity(name);
+
+    if (!(entity instanceof ConcreteEntity)) {
+        throw new Error(`Cannot add relationship to non-concrete entity ${entity.name}`);
+    }
+
+    const relationshipPropertyInterfaces = getRelationshipPropertiesInterfaces(definitionNodes);
+    if (!entity) throw new Error(`Entity ${name} not found while creating relationships`);
```
Comment (MacondoExpress) [inline packages/graphql/src/schema-model/generate-model.ts:147]: I would raise this error before!

**Review by MacondoExpress — APPROVED** [review]
I just left just nitpicking comments and a conversational point, no blocking points in my opinion.

File: packages/graphql/src/schema-model/generate-model.ts (hunk 3 above)
Comment (angrykoala) [reply packages/graphql/src/schema-model/generate-model.ts:147]: Good catch, in fact, this error is not needed, as `getEntity` already throws if none is found

File: packages/graphql/src/schema-model/relationship/Relationship.ts (hunk 1 above)
Comment (MacondoExpress) [reply packages/graphql/src/schema-model/relationship/Relationship.ts:37]: My idea is to have a set of classes that given a schema model generates the augmented schema and some stubs that can be used to collect these infos, it's just a rough idea after I saw this use case, we can discuss this in some  IR DDD sessions!

**Pull request merged**